### PR TITLE
[dev-v5] Add Drag & Drop component

### PR DESCRIPTION
**Pull request by MarvinKlein1508**
# Pull Request

## 📖 Description

This PR adds the `FluentDragContainer` and `FluentDropZone` components from v4 branch to v5. They are basically a copy & paste with just small changes to get them working in v5.

I also added the demo section and adjusted both test cases.

### 🎫 Issues

Implements #4208 

## 👩‍💻 Reviewer Notes

@vnbaaij & @dvoituron  not sure if you wanted to change something fundamental here in v5.

### General

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fluentui-blazor/blob/master/docs/contributing.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [x] I have added a new component
- [x] I have added [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for my new component
- [ ] I have modified an existing component
- [x] I have validated the [Unit Tests](https://github.com/Microsoft/fluentui-blazor/blob/master/unit-tests.md) for an existing component

## ⏭ Next Steps

I think in v5 we want to add test cases for each property right?

### CODE REVIEW

File: src/Core/Components/Drag/FluentDragContainer.razor [hunk 1]
```diff
@@ -0,0 +1,12 @@
+@namespace Microsoft.FluentUI.AspNetCore.Components
+@inherits FluentComponentBase
+
+@typeparam TItem
+@attribute [CascadingTypeParameter(nameof(TItem))]
+<div id=@Id
+     class=@ClassValue
+     style=@StyleValue>
```
Comment (vnbaaij) [inline src/Core/Components/Drag/FluentDragContainer.razor]: Must have  `@attributes="@AdditionalAttributes"`

File: src/Core/Components/Drag/FluentDragContainer.razor.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+public partial class FluentDragContainer<TItem> : FluentComponentBase
+{
+    /// <summary />
+    public FluentDragContainer(LibraryConfiguration configuration) : base(configuration) { }
+
+    /// <summary />
+    protected virtual string? ClassValue => new CssBuilder(Class).Build();
+
+    /// <summary />
+    protected virtual string? StyleValue => new StyleBuilder(Style).Build();
```
Comment (vnbaaij) [inline src/Core/Components/Drag/FluentDragContainer.razor.cs]: Must use `DefaultClassBuilder` and `DefaultStyleBuiler`

File: src/Core/Components/Drag/FluentDropZone.razor.cs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+public partial class FluentDropZone<TItem> : FluentComponentBase
+{
+    /// <summary />
+    public FluentDropZone(LibraryConfiguration configuration) : base(configuration) { }
+
+    /// <summary />
+    protected virtual string? ClassValue => new CssBuilder(Class).Build();
+
+    /// <summary />
+    protected virtual string? StyleValue => new StyleBuilder(Style).Build();
```
Comment (vnbaaij) [inline src/Core/Components/Drag/FluentDropZone.razor.cs]: Must use `DefaultClassBuilder` and `DefaultStyleBuiler`

File: src/Core/Components/Drag/FluentDropZone.razor [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+     @ondragstart:stopPropagation="@StopPropagation"
+     @ondragstart=@OnDragStartHandler
+     @ondragover=@OnDragOverHandler
+     @ondragover:preventDefault="@Droppable"
+     @ondragenter=@OnDragEnterHandler
+     @ondragenter:preventDefault="@Droppable"
+     @ondragleave=@OnDragLeaveHandler
+     @ondragleave:preventDefault="@Droppable"
+     @ondragend=@OnDragEndHandler
+     @ondrop=@OnDropHandler @ondrop:preventDefault>
```
Comment (vnbaaij) [inline src/Core/Components/Drag/FluentDropZone.razor]: Must have @attributes="@AdditionalAttributes"

**Review by vnbaaij — CHANGES_REQUESTED** [review]
The components must have @attributes in their razor files and use the v5 default class and style builders. With the way it is done now, unit tests are failing (because the standard parameters are not being utilized.

The components must be added in the `ComponentInitializer` dictionary in `ComponentBaseTests.cs` like this: 
```
{ typeof(FluentDragContainer<>), Loader.MakeGenericType(typeof(string))},
{ typeof(FluentDropZone<>), Loader.MakeGenericType(typeof(string))},
```

Also, a lot more tests need to be added. We are strict on having 98% line coverage for v5. Current result:

Image

So yes, all properties/parameters must have tests. And also yes, we didn;t have any additional plans for these components for V5.

**Comment by MarvinKlein1508** [comment]
@vnbaaij I have added your requested changes. Any idea why the ComponentBaseTest is failing?
grafik

**Comment by vnbaaij** [comment]
> @vnbaaij I have added your requested changes. Any idea why the ComponentBaseTest is failing?

You forgot to use default builders in drag container cs file

**Comment by MarvinKlein1508** [comment]
> > @vnbaaij I have added your requested changes. Any idea why the ComponentBaseTest is failing?
> 
> You forgot to use default builders in drag container cs file

oh wow. I forgot to save the file. Should be done now. I will add more UNIT tests later on. Can you tell me how I can see the coverage for the specific component?

Nevermind - the docs got me covered! 🤫

**Comment by MarvinKlein1508** [comment]
@vnbaaij I need your wisdom once again. This testing stuff is kinda new to me - so please show mercy! 🤣

How would you do it for the Draggable property? I guess I shouldn't declare two methods for true and false, right?

And how can I even check for mouse events? The docs shows how you can do it with `OnClick` but not with any other.

If you could give me like 1 or 2 examples then I will happily add the rest of the tests.

File: src/Core/Components/Drag/FluentDragEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    internal FluentDragEventArgs(FluentDropZone<TItem> source, FluentDropZone<TItem> target)
+    {
+        Source = source;
+        Target = target;
+    }
+
+    /// <summary>
+    /// Gets the zone where the drag started.
+    /// </summary>
+    public FluentDropZone<TItem> Source { get; } = null!;
```
Comment (dvoituron) [inline src/Core/Components/Drag/FluentDragEventArgs.cs]: Do not use `null!`, but use `required` instead if necessary.

File: src/Core/Components/Drag/FluentDropZone.razor.cs [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    protected virtual string? ClassValue => DefaultClassBuilder
+        .Build();
+
+    /// <summary />
+    protected virtual string? StyleValue => DefaultStyleBuilder
+        .Build();
+
+    /// <summary />
+    [CascadingParameter]
+    private FluentDragContainer<TItem> Container { get; set; } = default!;
```
Comment (dvoituron) [inline src/Core/Components/Drag/FluentDropZone.razor.cs]: Prefer to use `required` and not `default!`

File: src/Core/Components/Drag/FluentDropZone.razor.css [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿div[draggable='true'] {
+    cursor: move;
+}
+
+div[dragged-over] {
+    background-color: lightgray;
```
Comment (dvoituron) [inline src/Core/Components/Drag/FluentDropZone.razor.css]: Do we need to use a FluentUI CSS variable (and not always `lightgray`) ?

File: src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj [hunk 8]
```diff
@@ -125,7 +125,7 @@
     </ItemGroup>
 
     <!-- Log before modification -->
-    <Message Importance="normal" Text="Found @(_ContentWithAbsolutePaths->Count()) Content items with absolute paths to normalize" />
+    <Message Importance="normal" Text="Found @(_ContentWithAbsolutePaths-&gt;Count()) Content items with absolute paths to normalize" />
```
Comment (dvoituron) [inline src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj]: Why this change?

File: tests/Core/Components/Drag/FluentDragTests.razor [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿@using Microsoft.FluentUI.AspNetCore.Components.Tests.Extensions
+@using Microsoft.FluentUI.AspNetCore.Components.Utilities
+@using Xunit;
+@inherits Bunit.TestContext
+
+@code {
+	public FluentDragTests()
```
Comment (dvoituron) [inline tests/Core/Components/Drag/FluentDragTests.razor:8]: You will probably need to add additional unit tests to achieve at least 98% code coverage (100% if possible).

**Review by dvoituron — CHANGES_REQUESTED** [review]

File: src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj (hunk 8 above)
Comment (MarvinKlein1508) [reply src/Core/Microsoft.FluentUI.AspNetCore.Components.csproj]: Visual Studio doing Visual Studio things. I change it back.

File: src/Core/Components/Drag/FluentDropZone.razor.css (hunk 7 above)
Comment (MarvinKlein1508) [reply src/Core/Components/Drag/FluentDropZone.razor.css]: Good point. Which one would you recommend?

File: tests/Core/Components/Drag/FluentDragTests.razor (hunk 9 above)
Comment (MarvinKlein1508) [reply tests/Core/Components/Drag/FluentDragTests.razor:8]: Yea I know. Could you help me out by providing a testcase for the Draggable parameter? See: https://github.com/microsoft/fluentui-blazor/pull/4211#issuecomment-3371351243

**Comment by vnbaaij** [comment]
> @vnbaaij I need your wisdom once again. This testing stuff is kinda new to me - so please show mercy! 🤣
> 
> How would you do it for the Draggable property? I guess I shouldn't declare two methods for true and false, right?
> 
> And how can I even check for mouse events? The docs shows how you can do it with `OnClick` but not with any other.
> 
> If you could give me like 1 or 2 examples then I will happily add the rest of the tests.

I'd need to look into this on how to do it. Maybe you can draw inspiration from `FluentMultiSplitterTests`?

File: src/Core/Components/Drag/FluentDropZone.razor.css (hunk 7 above)
Comment (MarvinKlein1508) [reply src/Core/Components/Drag/FluentDropZone.razor.css]: @dvoituron @vnbaaij 

I've set it to
```css
background-color: var(--colorNeutralBackground2);
color: var(--colorNeutralForeground2);
```

any doubts?

Comment (dvoituron) [reply src/Core/Components/Drag/FluentDropZone.razor.css]: I can't approve the design at this time. But perhaps the colors “--colorNeutralBackground4” and 
“--colorNeutralForeground4” could be more visible? To check :-)

File: tests/Core/Components/Drag/FluentDragTests.razor (hunk 9 above)
Comment (dvoituron) [reply tests/Core/Components/Drag/FluentDragTests.razor:8]: I don't have time during these days. But I can check that later.
Perhaps you will find assistance by using GitHub Copilot.

File: src/Core/Components/Drag/FluentDropZone.razor.css (hunk 7 above)
Comment (MarvinKlein1508) [reply src/Core/Components/Drag/FluentDropZone.razor.css]: I've set them to 4 now.

**Comment by MarvinKlein1508** [comment]
I got some tests working with CoPilot which gets the coverage up to almost 100% for the three components. Only the FluentDropZone is now at 92.1%.

I can't figure out how I can satisfy the tests with the remaining edge cases like this:
grafik

grafik

grafik


In theory they starting zone should never be null. I believe the if statements are just here to prevent nullable warnings.

**Comment by MarvinKlein1508** [comment]
@vnbaaij @dvoituron this should be done now.

File: examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md [hunk 10]
```diff
@@ -0,0 +1,33 @@
+---
+title: Drag and Drop
+route: /Drag
+---
+
+# Drag and Drop
+
+A web component implementation of a <a href="https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API" target="_blank">HTML Drag and Drop API</a>.
```
Comment (dvoituron) [inline examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md]: A **Blazor** component...

File: examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+The user may select draggable elements with a mouse, drag those elements to a droppable element, and drop them by releasing the mouse button. A translucent representation of the draggable elements follows the pointer during the drag operation.
+
```
Comment (dvoituron) [inline examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md:11]: Add extra details:

Create a **FluentDragContainer** surface that will contain drag-and-drop areas. Customize these areas with the **Draggable** and **Droppable** parameters. Each time an item is dragged or dropped onto an area, an event is triggered to notify the developer. You can retrieve a generic **Item** argument with your personal data.

File: examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+* **Columns** can be moved within the same row or between different rows.
+* **Elements** can be rearranged inside a column or moved across columns.
+
+The <span style="background-color: yellow; padding: 0 0.2em; color: #000;">yellow area</span> indicates an empty drop zone inside a row. It accepts columns when a row does not contain any yet. Similarly, empty columns behave as drop zones for elements.
+
+This structure allows fully flexible layout editing with deep nesting and drag-and-drop handling at every level.
+
+{{ DragDropNested }}
+
+
```
Comment (dvoituron) [inline examples/Demo/FluentUI.Demo.Client/Documentation/Components/Drag/FluentDragDrop.md:44]: Add 
```
## API FluentDragContainer

{{ API Type=FluentDragContainer<int> }}

## API FluentDropZone

{{ API Type=FluentDropZone<int> }}

## API FluentDragEventArgs

{{ API Type=FluentDragEventArgs<int> }}
```
